[TFLite, 16x8] Quantization 16x8, fix for inference_input(output)_type

wwwind opened the pull request:
This is a small fix for the new options in TFLite converter to make them work with int16 activations:

`converter.inference_input_type = tf.int16`
`converter.inference_output_type = tf.int16`

wwwind committed:
Fix for inference_input(output)_type for 16x8.

Change-Id: I5791454da3c0ebb812c31e1ba304b745acc006e3

wwwind committed:
Addressed reviewer's comments.

Change-Id: Ida9f64eb66ee8e96840fdb84723286fccf0c2170

wwwind commented:
Hi @MeghnaNatraj Thanks for the review. I updated except for the test _testPostTrainingIntegerAllowFloatQuantization_
It uses default full integer conversion without setting target_spec.supported_ops and as result, activations are in int8.

When activations are supposed to be in int8, but inference_input_type is set to int16, what is the expected behaviour ?
1. error
2. quant/dequant nodes from int16 to int8 and back.
Thanks

wwwind committed:
Addressed reviewer's comment.

Change-Id: Iae72065849ac149c24ac3db35d5babc5616aa649

MeghnaNatraj commented:
It should throw as error, but we don't need a test for this. [For reference: int8 quantization supports uint8/int8 input/output types and int16x8 quantization supports int16 input/output types]

@wwwind For `testPostTrainingIntegerAllowFloatQuantization`, you can make an exact copy of your updated test `testPostTrainingIntegerNoFloatQuantization` and further make the following modifications:

0. [Avoid copy-paste error] Ensure you re-name it as `testPostTrainingIntegerAllowFloatQuantization`
1. Remove the `use_target_ops_flag` from the test.
2. 
Replace 
```
if use_target_ops_flag:
    quantized_converter.target_spec.supported_ops = [
        lite.OpsSet.TFLITE_BUILTINS_INT8
    ]
    if quantization_16x8:
      quantized_converter.target_spec.supported_ops = [
          lite.OpsSet.\
          EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8
      ]
    else:
      quantized_converter.target_spec.supported_ops = [
          lite.OpsSet.TFLITE_BUILTINS_INT8
      ]
  else:
    quantized_converter.target_spec.supported_types = [lite.constants.INT8]   
```

with 

```
if quantization_16x8:
  quantized_converter.target_spec.supported_ops = [
       lite.OpsSet.\
          EXPERIMENTAL_TFLITE_BUILTINS_ACTIVATIONS_INT16_WEIGHTS_INT8,
       lite.OpsSet.TFLITE_BUILTINS
      ]
```

wwwind committed:
Added error when we have mismatch in inference types + tests.

Change-Id: I8488748318be91f4367fc33aa8e7a01e4ff15729

wwwind commented:
Hi @MeghnaNatraj Thank you for the review!

I corrected the change so that now it will be an error if there is a mismatch in type of quantization and requested inference_input(output)_type.
I added a separate test for the raised error and for the case "allow float" as suggested.

MeghnaNatraj commented on tensorflow/lite/python/lite.py:
```diff
@@ -562,6 +568,11 @@ def _validate_inference_input_output_types(self, quant_mode):
         all_types_names = ["tf." + t.name for t in all_types]
         raise ValueError("The inference_input_type and inference_output_type "
                          "must be in {}.".format(all_types_names))
+    elif quant_mode.is_post_training_integer_quantize_16x8():
```
Replace this section with

```
elif quant_mode.is_post_training_integer_quantize_16x8():
      all_types = default_types + [constants.INT16]
      if self.inference_input_type not in all_types or \
          self.inference_output_type not in all_types:
        all_types_names = ["tf." + t.name for t in all_types]
        raise ValueError("The inference_input_type and inference_output_type "
                         "must be in {}.".format(all_types_names))
```

MeghnaNatraj reviewed (commented):
The changes look good - just one more modification is needed as i've added in the comments. After that I'll merge this PR.

wwwind committed:
Addressed reviewer's comments.

Change-Id: I99bdedc40dfdae8e6e6c57003bf1128051b40c1a

wwwind commented:
Hi @MeghnaNatraj Thanks for the review! Corrected.

MeghnaNatraj reviewed (approved):
Approving this.

Thank you for resolving all the comments so quickly! :)

MeghnaNatraj commented:
@gbaned  Feel free to merge this PR.
